### PR TITLE
Fixes #14369 - Removed default scope from host counter in taxonomies

### PR DESCRIPTION
**Pull request by ShimShtein**

### CODE REVIEW

**Comment by iNecas** [comment]
It shows the `Select Hosts` tab, but when I want to select the hosts manually, I get an empty table (probably due to same issue as the count)

**Comment by dLobatog** [comment]
Thanks for the fix! This is lacking tests, can you add some to ensure this bug doesn't happen again?

**Comment by ShimShtein** [comment]
Rebased and added tests.

**Comment by iNecas** [comment]
https://github.com/theforeman/foreman/pull/3371#issuecomment-202877487 is still not addressed

**Comment by ShimShtein** [comment]
> When I want to select the hosts manually, I get an empty table.

Right, I am not sure if it won't be a security breach - the user will be able to see hosts that do not belong to him.

**Comment by iNecas** [comment]
Well… he can just take their ownership right away without looking at them: but this is even for admin role: In my case, I'm logged in a admin, I can see the hosts, so I would expect I could select them in the manual assignment. I'm not proposing not using permissions there.

File: app/controllers/concerns/foreman/controller/taxonomies_controller.rb [hunk 1]
```diff
@@ -194,6 +194,6 @@ def find_resource
 
   def count_nil_hosts
     return @count_nil_hosts if @count_nil_hosts
-    @count_nil_hosts = Host.where(taxonomy_id => nil).count
+    @count_nil_hosts = Host.unscoped.where(taxonomy_id => nil).count
```
Comment (tbrisker) [inline app/controllers/concerns/foreman/controller/taxonomies_controller.rb]: `.authorized`? We don't to count hosts user isn't supposed to know about.

**Comment by ShimShtein** [comment]
@iNecas: Fixed you special use case - if you had an organization context and you tried to look for hosts without organization, the query always returned 0 results. It was due to the SQL that was generated: `where organization_id in (1) /* from the context */ and organization_id = nil /* from the no_organizations scope */`. I have changed those scopes to rewrite organization_id conditions by using `#rewhere` instead of usual `#where` that chains conditions.

@tbrisker: Same solution goes for your comment too - now I am overwriting only the taxonomy part, any other condition remains the same.

**Comment by ShimShtein** [comment]
Added permissions scope to all relevant methods.

File: test/unit/host_test.rb [hunk 2]
```diff
@@ -3130,6 +3130,28 @@ def to_managed!
...
 
+  describe 'taxonomy scopes' do
+    test 'no_location overrides default scope' do
+      location = FactoryGirl.create(:location)
+      host = FactoryGirl.create(:host, :location => nil)
+      Location.stubs(:current).returns(location)
+
+      assert_nil Host.where(:id => host.id).first
+      assert_not_nil Host.no_location.where(:id => host.id).first
+      Location.unstub(:current)
```
Comment (iNecas) [inline test/unit/host_test.rb:3141]: Do we need to unstub, when it's at the end of the test?

File: app/controllers/concerns/foreman/controller/taxonomies_controller.rb [hunk 3]
```diff
@@ -135,19 +135,19 @@ def import_mismatches
...
-    Host.send("no_#{taxonomy_single}").update_all(taxonomy_id => @taxonomy.id)
+    hosts_scope.send("no_#{taxonomy_single}").update_all(taxonomy_id => @taxonomy.id)
     @taxonomy.import_missing_ids
     redirect_to send("#{taxonomies_plural}_path"), :notice => _("All hosts previously with no %{single} are now assigned to %{name}") % { :single => taxonomy_single, :name => @taxonomy.name }
   end
 
   def assign_selected_hosts
     host_ids = params[taxonomy_single.to_sym][:host_ids] - ["0"]
-    @hosts = Host.where(:id => host_ids).update_all(taxonomy_id => @taxonomy.id)
+    @hosts = hosts_scope.where(:id => host_ids).update_all(taxonomy_id => @taxonomy.id)
```
Comment (iNecas) [inline app/controllers/concerns/foreman/controller/taxonomies_controller.rb]: `hosts_scope.send("no_#{taxonomy_single}")` was missing here, so the manual selection was not working properly. I would be ok fixing this as part of this PR, as the whole assignment was just broken.

**Comment by iNecas** [comment]
Some comments inline, but it's a step into right direction.

File: test/unit/host_test.rb (hunk 2 above)
Comment (ShimShtein) [inline test/unit/host_test.rb:3141]: For some reason if I don't unstub the call, it messes with other tests (at least on my PC). I think it has something to do with the fact that it's a static method, so the stub is not cleaned properly.

**Comment by ShimShtein** [comment]
Fixed `#assign_selected_hosts`

File: test/unit/host_test.rb (hunk 2 above)
Comment (iNecas) [inline test/unit/host_test.rb:3141]: aah, that might be it, thanks for explanation

**Comment by ShimShtein** [comment]
Test failure seems unrelated

**Comment by iNecas** [comment]
[test]

**Comment by iNecas** [comment]
I've tested it and works well. :+1:

**Comment by iNecas** [comment]
waiting for tests to finish

File: app/controllers/concerns/foreman/controller/taxonomies_controller.rb [hunk 4]
```diff
@@ -194,6 +194,16 @@ def find_resource
 
   def count_nil_hosts
     return @count_nil_hosts if @count_nil_hosts
-    @count_nil_hosts = Host.where(taxonomy_id => nil).count
+    @count_nil_hosts = hosts_scope.rewhere(taxonomy_id => nil).count
```
Comment (iNecas) [inline app/controllers/concerns/foreman/controller/taxonomies_controller.rb]: Could we use `hosts_scope_without_taxonomy` here?

**Comment by iNecas** [comment]
One suggestion added, but should be good to go after that

**Comment by ShimShtein** [comment]
Changed as suggested.

**Comment by iNecas** [comment]
Thanks @ShimShtein, merged as https://github.com/theforeman/foreman/commit/b597a8a5fea1e103af2d81b548b319ebdc92aa5c

**Pull request closed without merging**